compiler/rustc_hir_typeck/src/lib.rs

```diff
@@ -235,25 +236,25 @@ fn typeck_with_fallback<'tcx>(
         let param_env = tcx.param_env(def_id);
         let mut fcx = if let Some(hir::FnSig { header, decl, .. }) = fn_sig {
             let fn_sig = if rustc_hir_analysis::collect::get_infer_ret_ty(&decl.output).is_some() {
-                let fcx = FnCtxt::new(&inh, param_env, body.value.hir_id);
+                let def_id =
+                    inh.tcx.hir().opt_local_def_id(body.value.hir_id).unwrap_or(CRATE_DEF_ID);
```

Comment:
The `def_id` from the outer scope is the better value here: `body` is already the body owned by `def_id`.